beacon-chain/forkchoice/protoarray/optimistic_sync.go

```diff
@@ -2,321 +2,52 @@ package protoarray
-			return NonExistentNode, nil
-		}
-		node = s.nodes[node.parent]
-	}
-}
-
-// UpdateSyncedTipsWithValidRoot is called with the root of a block that was returned as
-// VALID by the EL. This routine recomputes and updates the synced_tips map to
-// account for this new tip.
-func (f *ForkChoice) UpdateSyncedTipsWithValidRoot(ctx context.Context, root [32]byte) error {
-	f.store.nodesLock.RLock()
-	defer f.store.nodesLock.RUnlock()
-	// We can only update if given root is in Fork Choice
-	index, ok := f.store.nodesIndices[root]
-	if !ok {
-		return errInvalidNodeIndex
-	}
-
-	// We can only update if root is a leaf in Fork Choice
-	node := f.store.nodes[index]
-	if node.bestChild != NonExistentNode {
-		return errInvalidBestChildIndex
-	}
-
-	// Stop early if the root is part of validated tips
-	f.syncedTips.Lock()
-	defer f.syncedTips.Unlock()
-	_, ok = f.syncedTips.validatedTips[root]
-	if ok {
-		return nil
-	}
-
-	// Cache root and slot to validated tips
-	newTips := make(map[[32]byte]types.Slot)
-	newValidSlot := node.slot
-	newTips[root] = newValidSlot
-
-	// Compute the full valid path from the given node to its previous synced tip
-	// This path will now consist of fully validated blocks. Notice that
-	// the previous tip may have been outside the Fork Choice store.
-	// In this case, only one block can be in syncedTips as the whole
-	// Fork Choice would be a descendant of this block.
-	validPath := make(map[uint64]bool)
-	validPath[index] = true
-	for {
-		if ctx.Err() != nil {
-			return ctx.Err()
-		}
-
-		parentIndex := node.parent
-		if parentIndex == NonExistentNode {
-			break
-		}
-		if parentIndex >= uint64(len(f.store.nodes)) {
-			return errInvalidNodeIndex
-		}
-		node = f.store.nodes[parentIndex]
-		_, ok = f.syncedTips.validatedTips[node.root]
-		if ok {
-			break
-		}
-		validPath[parentIndex] = true
-	}
-
-	// Retrieve the list of leaves in the Fork Choice
-	// These are all the nodes that have NonExistentNode as best child.
-	leaves, err := f.store.leaves()
-	if err != nil {
-		return err
-	}
-
-	// For each leaf, recompute the new tip.
-	for _, i := range leaves {
-		node = f.store.nodes[i]
-		j := i
-		for {
-			if ctx.Err() != nil {
-				return ctx.Err()
-			}
-			// Stop if we reached the previous tip
-			_, ok = f.syncedTips.validatedTips[node.root]
-			if ok {
-				newTips[node.root] = node.slot
-				break
-			}
-
-			// Stop if we reach valid path
-			_, ok = validPath[j]
-			if ok {
-				newTips[node.root] = node.slot
-				break
-			}
-
-			j = node.parent
-			if j == NonExistentNode {
+// pruneInvalid removes the node with the given root and all of its children
+// from the Fork Choice Store.
+func (s *Store) pruneInvalid(ctx context.Context, root [32]byte) error {
+	s.nodesLock.Lock()
+	defer s.nodesLock.Unlock()
+
+	node, ok := s.nodeByRoot[root]
+	if !ok || node == nil {
+		return errNilNode
+	}
+	if !node.optimistic || node.parent == nil {
+		return errInvalidOptimisticStatus
+	}
```

Comment:
We shouldn't return `errInvalidOptimisticStatus` when `node.parent == nil`.

Reply:
I think we do because the only situation when this can happen is when the finalized tree root is INVALID, and this is an invalid optimistic status. We should probably panic in the caller too.